### ISSUE TITLE
`ignore_text_within:` does not consider ancestor elements

### ISSUE THREAD

**Issue by ppworks** [gjtorikian/selma#85]
It seems that the `ignore_text_within:` option in `Selma::Selector` only checks the direct parent element, but does not account for ancestor elements. 

## Code example:

```ruby
require 'selma'
class MatchText
  SELECTOR = Selma::Selector.new(match_text_within: "*", ignore_text_within: ["code"])

  def selector
    SELECTOR
  end

  def handle_text_chunk(text)
    text.replace(text.to_s.sub('foo', 'bar'), as: :html)
  end
end

rewriter = Selma::Rewriter.new(sanitizer: nil, handlers: [MatchText.new])
result = rewriter.rewrite(<<~HTML)
<p>foo</p><code>foo<span class="highlight">foo</span></code>
HTML

puts result
```

### Expected behavior:
The text within `code` elements, including nested or wrapped content, should not be modified.
Expected output:

```
<p>bar</p><code>foo<span class="highlight">foo</span></code>
```

### Actual behavior:

The text within `code > span` element is modified, even though `ignore_text_within:` is set to exclude `code`. This suggests that only the direct parent is checked, rather than considering ancestor elements.
Actual output:

```
<p>bar</p><code>foo<span class="highlight">bar</span></code>
```

**Comment by gjtorikian** [comment]
Huh, interesting. I _think_ I know why this is happening, but, could you change your example to `code *` and try again? That might at least unblock you. (Sorry, it will be a few days before I can look into this properly!)

**Comment by ppworks** [comment]
> could you change your example to code * and try again?

I tried that, but it didn’t work. Thanks anyway!

```ruby
require 'selma'
class MatchText
  SELECTOR = Selma::Selector.new(match_text_within: "*", ignore_text_within: ["code *"])

  def selector
    SELECTOR
  end

  def handle_text_chunk(text)
    text.replace(text.to_s.sub('foo', 'bar'), as: :html)
  end
end

rewriter = Selma::Rewriter.new(sanitizer: nil, handlers: [MatchText.new])
result = rewriter.rewrite(<<~HTML)
<p>foo</p><code>foo<span class="highlight">foo</span></code>
HTML

puts result
```

## result

```
<p>bar</p><code>bar<span class="highlight">bar</span></code>
```

**Comment by gjtorikian** [comment]
Ok, thanks for trying that! I can get to this within the next few weeks. I wrote a test based on the bug and I am really sure I know what's going on. The downside is that I actually don't think the CSS parser supports hierarchy at all (meaning, `code span`, `code *`, etc, _should_ work but don't). It's a separate issue but I'm trying to ideate on how to work around that specifically.

**Comment by gjtorikian** [comment]
Well, the PaaS I use at work for deployments was down today, so I fixed this instead.

**Comment by ppworks** [comment]
Thank you so much for addressing the issue!
